text/XXXX-replace_publish_clone_with_push_pull.md

```diff
@@ -0,0 +1,160 @@
+can just as easily send to another peer that is configured to accept datasets 
+from you. For details on  setting peers for dataset exchange read this doc:
+https://qri.io/docs/hosting-datasets
+
+Usage:
+  qri push [REMOTE] DATASET [DATASET] [flags]
+
+Aliases:
+  publish
+
+Examples:
+  # push a dataset to the registry
+  $ qri push me/dataset
+
+  # push a specific version of a dataset to the registry:
+  $ qri push me/dataset@/ipfs/QmHashOfVersion
+
+  # push only the latest two versions to a peer named xiao
+  $ qri push xiao me/dataset --revisions 2
+
+Flags:
+  -l, --logs               send only dataset history
+  -r, --revisions          send only the lastest commit data
+```
+
+### pull: download datasets from the qri network
+```
+$ qri pull --help
+Pull downloads datasets and stores them locally, fetching the dataset log and
+dataset version(s). By default pull only fetches the latest version of a dataset
+
+By default pull fetches 
+
+Because pull grabs the latest version by default, it's common to 
+
+If qri connect is running
+
+Usage:
+  qri pull [REMOTE] [DATASET] [flags]
+
+Examples:
+  # download a dataset log and latest version
+  $ qri pull b5/world_bank_population
+
+  # download a dataset log and all versions
+  $ qri pull b5/world_bank_population --all
+
+  # pull a dataset from 
+  $ qri pull ramfox b5/world_bank_population
+
+Flags:
+  -l, --logs              save only dataset history
+  -r, --revisions         fetch a set number of versions from the latest commit
+  -a, --all               fetch all versions, overrides revisions
+  -k, --keep-files        when in a working directory, don't update files
+
+```
+
+`revisions` and `all` flags mimic the way `qri remove` works.
+
+### clone: pull and checkout a dataset
+
+We already have "clone" language in qri cloud, and have intended to rename `add` to `clone` for some time. The problem is add isn't the same thing as clone when compared to git. Removing the idea of publish & unpublish gives us back a little cognitive overhead that we can use for the classic combo of "get the latest version and checkout a directory
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+* history syncing 
+* dsync needs support for a manifest with multiple roots, multiple pins
+* receivers may not honor all block push or pin requests
+* pushing with `--latest` makes push behave like pull
+* pulling with `--all` makes pull behave like default push
+* All Peers are Potential Remotes
+* IPFS is the natural fallback network
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Push & Pull don't work like git
+This is a problem, but I think the deviation is justified because we want different things from dataset version control than code version control. Datasets can get big. Versions of datasets can get bigger.
+
+We need to make it _very_ clear in documentation that pulling a dataset doesn't get you all versions by default.
```

Comment:
One of the things I was thinking about as I read this was: I’m not *usually* concerned about the cost of pulling a Git repo☨, but I often would be when it comes to data. It’s just more likely to get bigger faster, and I’m *probably* less likely to need to investigate or validate the history. Do you think there might be a feasible way to estimate the cost of pulling N revisions before actually doing so? (It doesn’t seem easy since Qri versions are trees with lots of common nodes between them rather than diffs that build up.)


*☨: Obviously there are exceptions for some bigger projects, but those situations are rare for most people and usually made as reasonable as possible by maintainers splitting up repos or squashing histories every so often. There aren’t similar history manipulation tools in Qri, and I’d bet many potential publishers or registries would be philosophically against squashing like that (that kind of thing feels somehow more sacred with *data* than a work product or tool like code).*

Reply:
I like this a lot. It also fits nicely in the story of the SQL query prompt being able to show the size of data that will be fetched/scanned so that the query can be executed. Also think this would be useful for adjusting all sorts of behaviors, or at least have a simple guard for users like "hey you're going to download 1TB of data, are you sure (y/N)?"

Reply:
into it. I'm going to add it the the "unresolved questions" section.